File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if not needle:
            return 0

        i, j = 0, 0
        hay_len = len(haystack)
        needle_len = len(needle)

        while i + j < hay_len:
            if haystack[i + j] == needle[j]:
                if j == needle_len - 1:
                    return i
                j += 1
            else:
                i += 1
                if j > 0 and haystack[i:i + j] == needle[:j]:
                    j -= 1
                else:
                    j = 0

        return -1

File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.c
int strStr(char* haystack, char* needle) {

    int i = 0;
    int j = 0;
    int hayLen = strlen(haystack);
    int needleLen = strlen(needle);

    while (i + j < hayLen) {
        if (haystack[i + j] == needle[j]) {  
            if (j == needleLen - 1) return i;  
            j++;
        } else {
            i++;

            if (j > 0 && strncmp(haystack + i, needle, j) == 0)
                j--;
            else
                j = 0;
        }
    }
    return -1;  // No match found
}
